db/keyval/etcd/plugin_impl_etcd.go

```diff
@@ -154,18 +163,99 @@ func (plugin *Plugin) Compact(rev ...int64) (toRev int64, err error) {
 	return 0, fmt.Errorf("connection is not established")
 }
 
-func (plugin *Plugin) getEtcdConfig() (Config, error) {
+// Method starts loop which attempt to connect to the ETCD. If successful, send signal callback with resync,
+// which will be started when datasync confirms successful registration
+func (plugin *Plugin) etcdReconnectionLoop(clientCfg *ClientConfig) {
+	var err error
+	// Set reconnect interval
+	interval := plugin.config.ReconnectInterval
+	if interval == 0 {
+		interval = defaultReconnectInterval
+	}
+	plugin.Log.Infof("ETCD server %s not reachable in init phase. Agent will continue to try to connect every %d second(s)",
+		plugin.config.Endpoints, interval)
+	for {
+		time.Sleep(time.Duration(interval) * time.Second)
+		plugin.Log.Infof("Connecting to ETCD %v ...", plugin.config.Endpoints)
+		plugin.connection, err = NewEtcdConnectionWithBytes(*clientCfg, plugin.Log)
+		if err != nil {
+			continue
+		} else {
+			plugin.Log.Infof("ETCD server %s connected", plugin.config.Endpoints)
+			// Configure connection and set as connected
+			plugin.configureConnection()
+			plugin.connected = true
+			// Execute callback functions
+			for _, callback := range plugin.onConnection {
+				if err := callback(); err != nil {
+					plugin.Log.Error(err)
+				}
+			}
+			// At the end, call resync
+			if plugin.Resync != nil {
+				plugin.Resync.DoResync()
+			}
+			plugin.Log.Debugf("Etcd reconnection loop ended")
+			return
+		}
+	}
+}
+
+// If ETCD is connected, complete all other procedures
+func (plugin *Plugin) configureConnection() {
+	plugin.reconnectResync = plugin.config.ReconnectResync
+	if plugin.config.AutoCompact > 0 {
+		if plugin.config.AutoCompact < time.Duration(time.Minute*60) {
+			plugin.Log.Warnf("Auto compact option for ETCD is set to less than 60 minutes!")
+		}
+		plugin.startPeriodicAutoCompact(plugin.config.AutoCompact)
+	}
+	plugin.protoWrapper = kvproto.NewProtoWrapperWithSerializer(plugin.connection, &keyval.SerializerJSON{})
+}
+
+// Starts periodic ETCD status check
+func (plugin *Plugin) startStatusCheck() {
```

Comment:
This is way too nested. It will be much more readable to do the statuscheck registration in `Init` and use this function as `probe` argument for the `Register`.